src/cowboy_http2.erl

```diff
@@ -938,6 +964,54 @@ goaway_streams(State, [{StreamID, #stream{state=StreamState}}|Tail], LastStreamI
 goaway_streams(State, [Stream|Tail], LastStreamID, Reason, Acc) ->
 	goaway_streams(State, Tail, LastStreamID, Reason, [Stream|Acc]).
 
+%% A server that is attempting to gracefully shut down a connection SHOULD send
+%% an initial GOAWAY frame with the last stream identifier set to 2^31-1 and a
+%% NO_ERROR code. This signals to the client that a shutdown is imminent and
+%% that initiating further requests is prohibited. After allowing time for any
+%% in-flight stream creation (at least one round-trip time), the server can send
+%% another GOAWAY frame with an updated last stream identifier. This ensures
+%% that a connection can be cleanly shut down without losing requests.
```

Comment:
I think you should add a `closing_initiated` state that becomes `closing` when the timeout triggers. The difference between the two would be that we don't terminate the connection if all streams finish in `closing_initiated`.

Reply:
OK.

A few guards checking for `Status =:= connected` will have to be changed to `Status =:= connected; Status =:= closing_initiated` but that's fine. If `sys:terminate/2,3` is called while in `closing_initiated`, it can be ignored.

I reused the `connected` state instead of introducing a new one `closing_initiated` since there isn't really any logic which needs to know the difference, but I guess the scenario where `sys:terminate/2,3` is called twice adds enough motivation for a new state, plus it's more clear to have a different state.